### PR TITLE
fix use-after-free and buffer-overflow found with AddressSanitizer ru…

### PR DESCRIPTION
**Pull request by AntonioBL**
…nning over the tests

### CODE REVIEW

**Comment by Jojo-Schmitz** [comment]
Came up in https://musescore.org/en/node/276356#comment-859719,along with a patch to fix some memory leaks: [memory_leak_diff.txt](https://github.com/musescore/MuseScore/files/2534587/memory_leak_diff.txt)

File: libmscore/score.cpp [hunk 1]
```diff
@@ -315,6 +315,10 @@ Score::~Score()
       // deselectAll();
       for (MeasureBase* m = _measures.first(); m;) {
             MeasureBase* nm = m->next();
+            if (m->system()) {
+                  auto i = std::find(m->system()->measures().begin(), m->system()->measures().end(), m);
+                  m->system()->measures().erase(i);
+                  }
```
Comment (mirabilos) [inline libmscore/score.cpp]: Oooooh, this will fix (some of) the crashes I was seeing!

+1! +1! +1!

File: libmscore/spanner.cpp [hunk 2]
```diff
@@ -304,6 +304,12 @@ Spanner::Spanner(const Spanner& s)
 
 Spanner::~Spanner()
       {
+      for (SpannerSegment* ss : segments)
+            if (ss->system())
+                  ss->system()->remove(ss);
+      for (SpannerSegment* ss : unusedSegments)
+            if (ss->system())
+                  ss->system()->remove(ss);
```
Comment (mirabilos) [inline libmscore/spanner.cpp]: This smells like another crash I was seeing…

**Comment by mirabilos** [comment]
+100! This is wonderful! With this, I can restore the `delete score;` in the batch job conversion, and the only remaining Valgrind warning I get (after having converted 52 scores to 209 output files in one go) is:

    Conditional jump or move depends on uninitialised value(s)
       at 0xBA46AB: Ms::TextLineBaseSegment::layout() (in /usr/bin/mscore-snapshot)
       by 0xAB568E: Ms::HairpinSegment::layout() (in /usr/bin/mscore-snapshot)
       by 0xAEFDD3: Ms::SLine::layoutSystem(Ms::System*) (in /usr/bin/mscore-snapshot)
       by 0xADDA07: ??? (in /usr/bin/mscore-snapshot)
       by 0xAE92E4: Ms::Score::collectSystem(Ms::LayoutContext&) (in /usr/bin/mscore-snapshot)
       by 0xAEAE36: Ms::LayoutContext::collectPage() (in /usr/bin/mscore-snapshot)
       by 0xAEBF09: Ms::LayoutContext::layout() (in /usr/bin/mscore-snapshot)
       by 0xAEC854: Ms::Score::doLayoutRange(int, int) (in /usr/bin/mscore-snapshot)
       by 0xBDCE06: Ms::Score::update() (in /usr/bin/mscore-snapshot)
       by 0x818B81: Ms::readScore(Ms::MasterScore*, QString, bool) (in /usr/bin/mscore-snapshot)
       by 0x81925C: Ms::MuseScore::readScore(QString const&) (in /usr/bin/mscore-snapshot)
       by 0x686422: ??? (in /usr/bin/mscore-snapshot)

Also:

    LEAK SUMMARY:
       definitely lost: 1,251,473 bytes in 4,826 blocks
       indirectly lost: 6,287,212 bytes in 46,222 blocks
         possibly lost: 2,434,028 bytes in 30,888 blocks

File: libmscore/spanner.cpp (hunk 2 above)
Comment (MarcSabatella) [reply libmscore/spanner.cpp]: Me too, crash in SpannerSegment destructor on closing a score

File: libmscore/xmlreader.cpp [hunk 3]
```diff
@@ -604,6 +604,8 @@ void XmlReader::removeConnector(const ConnectorInfoReader& cref)
             c = c->prev();
       while (c) {
             ConnectorInfoReader* next = c->next();
+            if (next)
+                  next->resetPrev();
```
Comment (dmitrio95) [inline libmscore/xmlreader.cpp]: This change should fix the use-after-free issue but there seem to be other issues with connector readers, see [my commit](https://github.com/musescore/MuseScore/pull/4278/commits/76a22bd8315b9bfa884d5e1284eb2a16494f691f) option to fix this.

File: libmscore/spanner.cpp (hunk 2 above)
Comment (dmitrio95) [reply libmscore/spanner.cpp]: Concerning these changes (this and the one in Score destructor) it is also possible to fix this by making systems be deleted before other elements: https://github.com/musescore/MuseScore/pull/4278/commits/e85f3a218cd275f003504044ce73bca3416be23c. Not sure which approach is better though.

File: libmscore/xmlreader.cpp (hunk 3 above)
Comment (AntonioBL) [reply libmscore/xmlreader.cpp]: I was simply trying a minimal change to cope with the bug: unfortunately, I really do not understand in deep detail the code of the connector readers.

File: libmscore/spanner.cpp (hunk 2 above)
Comment (AntonioBL) [reply libmscore/spanner.cpp]: In any case, we have to make sure that deleting the system does not try to delete objects belonging to the score.
I also don't know which approach would be the better.
The important thing is to fix bug, somehow :-)

**Comment by AntonioBL** [comment]
@mirabilos : yes, there are actually a lot of small memory leaks, which when processing a certain number of score pile up. Luckily, the worst memory leaks (which are causing memory exhaustion in version 2.3.2) have already been taken care of.

**Comment by mirabilos** [comment]
> Seems Travis CI failed for unrelated reasons, guess a restart would succeed

Yes, the code builds fine and helps my other PR along.

**Comment by anatoly-os** [comment]
See #4278

**Pull request closed without merging**